src/goto-instrument/code_contracts.cpp

```diff
@@ -244,6 +245,109 @@ void code_contractst::add_quantified_variable(
   }
 }
 
+optionalt<exprt> code_contractst::replace_old_parameter(
+  exprt expr,
+  std::map<exprt, exprt> &parameter2history,
+  source_locationt location,
+  const irep_idt &function,
+  const irep_idt &mode,
+  std::list<goto_programt::instructiont> &instructions)
+{
+  bool modified = false;
+
+  for(auto &op : expr.operands())
+  {
+    auto op_result = replace_old_parameter(
+      op, parameter2history, location, function, mode, instructions);
+    if(op_result.has_value())
+    {
+      op = *op_result;
+      modified = true;
+    }
+  }
+
+  if(expr.id() == ID_old)
+  {
+    DATA_INVARIANT(expr.operands().size() == 1, "old must have one operands");
+
+    const auto &parameter = to_old_expr(expr).variable();
+
+    // TODO: generalize below
+    PRECONDITION(parameter.id() == ID_dereference);
+
+    exprt ret;
+    if(parameter.id() == ID_dereference)
+    {
+      const auto &dereference_expr = to_dereference_expr(parameter);
+      const auto &cur_pointer = dereference_expr.pointer();
+
+      std::map<exprt, exprt>::iterator it = parameter2history.find(cur_pointer);
+
+      if(it == parameter2history.end())
+      {
+        // 1. Create a temporary symbol expression that represents the
+        // history variable
+        symbol_exprt tmp_symbol =
+          new_tmp_symbol(cur_pointer.type(), location, function, mode)
+            .symbol_expr();
+
+        // 2. Associate the above temporary variable to it's corresponding
+        // expression
+        parameter2history[cur_pointer] = tmp_symbol;
+
+        // 3. Add the required instructions to the instructions list
+        // 3.1 Declare the newly created temporary variable
+        instructions.push_back(goto_programt::make_decl(tmp_symbol, location));
+
+        // 3.2 Add an assignment such that the value pointed to by the new
+        // temporary variable is equal to the value of the corresponding
+        // parameter
+        instructions.push_back(goto_programt::make_assignment(
+          dereference_exprt(tmp_symbol),
+          dereference_exprt(cur_pointer),
+          location));
+      }
+
+      ret = dereference_exprt(parameter2history[cur_pointer]);
+    }
+
+    return ret;
+  }
+  else if(modified)
+    return std::move(expr);
+  else
+    return {};
+}
+
+std::pair<goto_programt, std::list<goto_programt::instructiont>>
+code_contractst::create_ensures_instruction(
+  codet expression,
```

Comment:
I was able to change it to `codet &expression`. However I did make some slight modifications inside the function and at the call sites. But overally, I believeI was able to provide a certain improvement.

Reply:
Nice.

I can review the change when after you push it. I'm curious why `const codet &` wouldn't work:
- if you're modifying `expression` within this function, then `codet expression` would have been incorrect in the first place, because that was a copy
- if you're not modifying `expression` within this function, then `const codet &` should work fine